snuba/query/matcher.py

```diff
@@ -0,0 +1,218 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from datetime import date, datetime
+from typing import Generic, Mapping, Optional, Sequence, Tuple, TypeVar
+
+from snuba.query.expressions import Column as ColumnExpr
+from snuba.query.expressions import Expression, FunctionCall
+
+TNode = TypeVar("TNode")
+
+
+@dataclass(frozen=True)
+class MatchResult:
+    """
+    Contains the parameters found in an expression by a Pattern.
+    Parameters are the equivalent to groups in regular expressions, they
+    are named parts of the expression we are matching that are identified
+    when a valid match is found.
+    """
+
+    expressions: Mapping[str, Expression] = field(default_factory=dict)
+    strings: Mapping[str, Optional[str]] = field(default_factory=dict)
+    ints: Mapping[str, Optional[int]] = field(default_factory=dict)
+    floats: Mapping[str, Optional[float]] = field(default_factory=dict)
+    dates: Mapping[str, Optional[date]] = field(default_factory=dict)
+    datetimes: Mapping[str, Optional[datetime]] = field(default_factory=dict)
```

Comment:
In theory yes, the optionals are intentionsl. There are places where I would make sense to explicitly match a None. An example would be to match a Column that has alias == None.
the String pattern is capable of matching None. Though you are rightm the isinsttance would not work properly in that case. I think I will remove the optional from the parameters and allow you to match None through a Pattern.

Expression was intentionally not Optional because there is no case where you want to match an Expression that could be None. It cannot happen in the AST.